R/oligoNucleotideEnrichment.R

```diff
@@ -0,0 +1,188 @@
+                                      alpha=0.05){
+  if(!file.exists(filepath)){
+    stop("file doesn't exist!")
+  }
+  background <- match.arg(background)
+  method <- match.arg(method)
+  if(!is(genome, "BSgenome") & !is(genome, "Mart") | missing(genome)){
+    stop("genome is required parameter, 
+             please pass in either a BSgenome object or a Mart object!")
+  }
+  if(round(times*alpha,0)==0){
+    stop("The paramter 'times' should be increased to large enough!")
+  }
+  nPeak <- length(peaks)
+  allseqLen <- seqlengths(genome)
+  seqWidth <- width(peaks)
+  candiChrom <- names(allseqLen >= max(seqWidth))
+  if(any(seqWidth > allseqLen[as.character(seqnames(peaks)@values)])){
+    stop("The length of sequence should be less than the length of chromosomes")
+  }
+  actualSeqs <- getAllPeakSequence(peaks, upstream = 0, downstream = 0, genome = genome)
+  actualP <- oligoNucleotideSummary(filepath,format=format,seqs=actualSeqs)
+  
+  statDist <- matrix(NA,0,4)
+  colnames(statDist) <- c("x","n","prop.background","binom.pvalue")
+  if(background=="randomPick"){
+    statDist <- do.call(rbind,lapply(seq.int(times),function(n) {
+      if(is.null(chromosome)==TRUE){
+        chrs <- as.character(seqnames(peaks))
+        givenSeqLen <- allseqLen[as.character(seqnames(peaks))]
+      }else{
+        chrs <- sample(candiChrom,nPeak)
+        givenSeqLen <- allseqLen[chrs]
+      }
+      startPos <- sapply(givenSeqLen-seqWidth,function(x){sample(seq.int(x),1)})
+      endPos <- startPos + seqWidth - 1
+      backPeak <- GRanges(seqnames=chrs,
+                          IRanges(start=startPos,
+                                  end=endPos,
+                                  names=paste0("peak",seq.int(nPeak))))
+      backSeq <- getAllPeakSequence(backPeak, upstream = 0, downstream = 0, genome = genome)
+      statTmp <- oligoNucleotideSummary(filepath,format=format,seqs=backSeq)
+      rownames(statTmp) <- paste0(n,"_",rownames(statTmp))
+      statTmp
+    }))
+  }else if(background=="shuffle"){
+    colnames(mcols(actualSeqs))[3] <- "orig.sequence"
+    names(actualSeqs$orig.sequence) <- do.call(paste, c(as.data.frame(actualSeqs)[,1:3], sep="_"))
+    inputSeq <- DNAStringSet(actualSeqs$orig.sequence)
+    statDist <- do.call(rbind,lapply(seq.int(times),function(n){
+      sequences.shuffled <- universalmotif::shuffle_sequences(inputSeq, ...)
+      actualSeqs$sequence <- as.vector(sequences.shuffled)
+      statTmp <- oligoNucleotideSummary(filepath,format=format,seqs=actualSeqs)
+      rownames(statTmp) <- paste0(n,"_",rownames(statTmp))
+      statTmp
+    }))
+  }
+  sortStatDist <- do.call(cbind,lapply(seq.int(nrow(actualP)),function(i){
+    sort(as.numeric(statDist[seq(i,nrow(statDist),nrow(actualP)),4]))
+  }))
+  colnames(sortStatDist) <- seq.int(nrow(actualP))
+  thresholdP <- sortStatDist[round(times*alpha,0),]
+  actualP <- cbind(actualP,thresholdP)
+  colnames(actualP)[5] <- "threshold"
```

Comment:
check dimension before assign colnames.

Reply:
actualP is a pre-defined dataframe/matrix with 4 columns.